### PR TITLE
Port VSTest to FAKE 5

### PR DESCRIPTION
**Pull request by DanielFabian**
I copied the MSTest module and made the appropriate changes for VSTest. Please take a first look at the code.

So far, I haven't been able to test it. To do so, I understand, I need to build the nuget locally. Do I have to run the `build.fsx` with some particular target or something?

### CODE REVIEW

File: src/app/Fake.DotNet.Testing.VSTest/VSTest.fs [hunk 1]
```diff
@@ -0,0 +1,150 @@
+﻿/// Contains tasks to run [VSTest](https://msdn.microsoft.com/en-us/library/ms182486.aspx) unit tests.
+module Fake.DotNet.Testing.VSTest
```
Comment (matthid) [inline src/app/Fake.DotNet.Testing.VSTest/VSTest.fs]: please add `[<RequireQualifiedAccess>]`

File: src/app/Fake.DotNet.Testing.VSTest/VSTest.fs [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+
+open Fake.Core
+open Fake.Testing.Common
+open System
+open System.Text
+
+/// [omit]
+let vsTestPaths =
```
Comment (matthid) [inline src/app/Fake.DotNet.Testing.VSTest/VSTest.fs]: internal?

File: src/app/Fake.DotNet.Testing.VSTest/VSTest.fs [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+/// [omit]
+let vsTestExe = 
+    if Environment.isMono then failwith "VSTest is not supported on the mono platform"
+    else "vstest.console.exe"
+
+/// Option which allow to specify if a VSTest error should break the build.
+type ErrorLevel = TestRunnerErrorLevel
+
+/// Parameter type to configure [VSTest.Console.exe](https://msdn.microsoft.com/en-us/library/jj155800.aspx)
+[<CLIMutable>]
```
Comment (matthid) [inline src/app/Fake.DotNet.Testing.VSTest/VSTest.fs]: remove CliMutable

File: src/app/Fake.DotNet.Testing.VSTest/VSTest.fs [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+      WorkingDir : string
+      /// A timeout for the test runner (optional).
+      TimeOut : TimeSpan
+      /// Error level for controlling how VSTest failures should break the build (optional).
+      ErrorLevel : ErrorLevel 
+      /// Path to test adapter e.g. xUnit (optional)
+      TestAdapterPath: string}
+
+/// VSTest default parameters.
+let VSTestDefaults =
```
Comment (matthid) [inline src/app/Fake.DotNet.Testing.VSTest/VSTest.fs]: internal

File: src/app/Fake.DotNet.Testing.VSTest/VSTest.fs [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+          | Some path -> path
+          | None -> ""
+      WorkingDir = null
+      TimeOut = TimeSpan.MaxValue
+      ErrorLevel = ErrorLevel.Error
+      TestAdapterPath = null }
+
+/// Builds the command line arguments from the given parameter record and the given assemblies.
+/// [omit]
+let buildVSTestArgs (parameters : VSTestParams) assembly =
```
Comment (matthid) [inline src/app/Fake.DotNet.Testing.VSTest/VSTest.fs]: internal?

File: src/app/Fake.DotNet.Testing.VSTest/VSTest.fs [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+///  - `setParams` - Function used to manipulate the default VSTestParams values.
+///  - `assemblies` - Sequence of one or more assemblies containing Microsoft Visual Studio Unit Test Framework unit tests.
+/// 
+/// ## Sample usage
+///
+///     Target.create "Test" (fun _ ->
+///         !! (testDir + @"\*.Tests.dll") 
+///           |> VSTest.VSTest (fun p -> { p with SettingsPath = "Local.RunSettings" })
+///     )
+let VSTest (setParams : VSTestParams -> VSTestParams) (assemblies : string seq) =
```
Comment (matthid) [inline src/app/Fake.DotNet.Testing.VSTest/VSTest.fs]: Maybe rename that such that it looks like `VSTest.run` for example (look how other runners call it).

**Comment by DanielFabian** [comment]
I made the suggested changes and I propose adding refactoring the remaining test runners into the backlog, because none of the following are currently in line with these changes: 

- [ ] `MSpec`
- [ ] `MSTest`
- [ ] `NUnit`
- [ ] `OpenCover`
- [ ] `XUnit`

**Comment by matthid** [comment]
Still missing is the updates for the warnings in https://github.com/fsharp/FAKE/blob/release/next/src/legacy/FakeLib/UnitTest/VSTest.fs to point to the new module. And ideally an update to the documentation template to add it to me menu (https://github.com/fsharp/FAKE/blob/release/next/help/templates/template.cshtml).

Thanks!

> So far, I haven't been able to test it. To do so, I understand, I need to build the nuget locally. Do I have to run the build.fsx with some particular target or something?

That is one way, and yes it should be generated once the new module is added to the `.sln` and to the list in the `build.fsx` which you have done both already afaics. A .nupkg should have been created for you...

**Comment by DanielFabian** [comment]
I don't quite know what isn't working with the CI... on my machine the build.fsx was green. Any idea?

**Comment by matthid** [comment]
Maybe our fake-templates broke, lets see if restarting works.

**Comment by matthid** [comment]
Thanks looks good! We figured out the test-error.
